kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java

```diff
@@ -126,33 +116,56 @@ private void processKieService(ClassLoader classLoader, String key, String value
                     childServices.computeIfAbsent( serviceName, k -> new ArrayList<>() )
                             .add( newInstance( classLoader, value.substring( 1 ) ) );
                 } else {
-                    services.put( serviceName, newInstance( classLoader, value ) );
+                    int priority = 0;
+                    int separatorPos = value.indexOf( ';' );
+                    if (separatorPos > 0) {
+                        priority = Integer.parseInt( value.substring( separatorPos+1 ).trim() );
+                        value = value.substring( 0, separatorPos ).trim();
+                    }
+                    services.put( priority, serviceName, newInstance( classLoader, value ) );
+                    log.debug( "Added Service " + value + " with priority " + priority );
                 }
             } catch (RuntimeException e) {
                 if (optional) {
                     log.info("Cannot load service: " + serviceName);
                 } else {
-                    System.out.println("Loading failed because " + e.getMessage());
+                    log.error("Loading failed because " + e.getMessage());
                     throw e;
                 }
             }
-            log.debug( "Adding Service {}\n", value );
         }
     }
 
-    @FunctionalInterface
-    private interface ServiceProcessor {
-        boolean process(ClassLoader classLoader, String key, String value);
-    }
-
     private <T> T newInstance( ClassLoader classLoader, String className ) {
         try {
-            return (T) Class.forName( className, true, classLoader ).newInstance();
+            return (T) Class.forName( className, true, classLoader ).getConstructor().newInstance();
         } catch (Throwable t) {
             throw new RuntimeException( "Cannot create instance of class: " + className, t );
         }
     }
 
+    private Map<String, List<Object>> buildMap() {
+        Map<String, List<Object>> servicesMap = new HashMap<>();
+        for (Map.Entry<String, List<Object>> serviceEntry : services.entrySet()) {
+            log.debug( "Service " + serviceEntry.getKey() + " is implemented by " + serviceEntry.getValue().get(0) );
+            servicesMap.put(serviceEntry.getKey(), serviceEntry.getValue());
+            List<?> children = childServices.remove( serviceEntry.getKey() );
+            if (children != null) {
+                for (Object child : children) {
+                    for (Object service : serviceEntry.getValue()) {
+                        (( Consumer ) service).accept( child );
+                    }
+                }
+            }
+        }
+
+        if (!childServices.isEmpty()) {
+            throw new RuntimeException("Child services " + childServices.keySet() + " have no parent");
+        }
+
+        return servicesMap;
```

Comment:
Can you please log as `debug` or `trace` the whole content of the map (with one or more lines of log)? So that it will be easy to get the full status of priorities and services.
It is fine for me also inside `getServices` method

Reply:
+1

Reply:
nevermind my previous comment, I understand is performed by https://github.com/kiegroup/droolsjbpm-knowledge/pull/466/files#diff-adbb6a46fc107d7f6590434788b73ac5R164-R172 as projected on each service. Thanks!